### PR TITLE
Optimize Random.nextString

### PR DESCRIPTION
**Pull request by joshlemer**
This implementation results in a 5x-10x speedup of the method.

### CODE REVIEW

File: src/library/scala/util/Random.scala [hunk 1]
```diff
@@ -188,14 +188,20 @@ class Random(val self: java.util.Random) extends AnyRef with Serializable {
...
-    def safeChar() = {
+  def nextString(length: Int): String = {
+    def safeChar(): Char = {
       val surrogateStart: Int = 0xD800
       val res = nextInt(surrogateStart - 1) + 1
       res.toChar
     }
-
-    List.fill(length)(safeChar()).mkString
+    val _length = java.lang.Math.max(length, 0)
```
Comment (retronym) [inline src/library/scala/util/Random.scala]: Could just return `""` when `n <= 0`.

**Comment by retronym** [comment]
You've accidentally checked in a `.DS_Store` file.

I recommend a global ignore of those:

```
➜  ~ git config --global core.excludesfile
~/.gitignore
➜  ~ cat ~/.gitignore | grep DS_Store
.DS_Store
```

File: src/library/scala/util/Random.scala [hunk 2]
```diff
@@ -188,14 +188,20 @@ class Random(val self: java.util.Random) extends AnyRef with Serializable {
...
+    val arr = new Array[Char](_length)
+    var i = 0
+    while (i < length) {
+      arr(i) = safeChar()
+      i += 1
+    }
```
Comment (retronym) [inline src/library/scala/util/Random.scala:207]: I'm happy to use the while loop here as it be slightly faster than `Array.fill(n)` which is generic and incurs some overhead.

Comment (joshlemer) [reply src/library/scala/util/Random.scala:207]: 😄

**Comment by diesalbla** [comment]
Note that the String constructor from Java creates another copy of the `Array[Char]`, so the one created in this method is to be discarded.

**Review by diesalbla — APPROVED** [review]

**Comment by joshlemer** [comment]
> Note that the String constructor from Java creates another copy of the `Array[Char]`, so the one created in this method is to be discarded.

Yeah, sadly there's absolutely no way to make a String without copying.

File: src/library/scala/util/Random.scala (hunk 2 above)
Comment (szeiger) [reply src/library/scala/util/Random.scala:207]: How about `Arrays.fill`?

**Pull request merged**